### ISSUE TITLE
no-constant-condition: new option to allow explicit true/false

### ISSUE THREAD

**Issue by guymguym** [eslint/eslint#10081]
**Tell us about your environment**

* **ESLint Version:** 4.14.0
* **Node Version:** 8.9.4
* **npm Version:** 5.6

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
module.exports = {

    // prevent from looking up other config files in containing folders
    root: true,

    // our environment is node.js with it's es6 extensions
    env: {
        node: true,
        es6: true,
    },

    parserOptions: {
        ecmaVersion: 2017
    },

    // RULES METHODOLOGY:
    // -----------------
    // we turn ON all of eslint rules by extending eslint:all
    // we selectively override the rules we want to remove or change.
    // See http://eslint.org/docs/user-guide/configuring#using-eslintall
    extends: 'eslint:all',

    rules: {

       // too long to paste

    }
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

We would like to allow constant conditions only if they are simply `true` or `false` because these are common patterns that we use for code with retries or complex conditions. 

We are aware that there are other ways to write retries logic which do not include constant conditions, but we believe this is going against the natural thinking process, and in most cases, this requires checking the retries conditions more than once.

We do not want to disable that rule because it protects us from unintended code such as `while (x=1)` and similar error prone patterns.

Example 1 - retries logic:
```js
while (true) {
  try {
    // do something that may throw
  } catch (err) {
    retries += 1;
    if (retries >= MAX_RETRIES) throw err;
  }
}
```

Example 2 - complex conditions:
```js
while (true) {
  const r1 = complex_condition_1();
  const r2 = complex_condition_2( r1 );
  const r3 = complex_condition_3( r1, r2 );
  // etc... you get the point.
  if (r42) break;
  // keep up doing stuff
}
```

```bash
eslint src/
```

**What did you expect to happen?**

We would like to suggest a new option to the `no-constant-condition` rule that will allow explicit `true` or `false` conditions.

**What actually happened? Please include the actual, raw output from ESLint.**

The rule gives an error and has no option to allow explicit true/false conditions.

**Comment by guymguym** [comment]
@dannyzaken FYI

**Comment by VictorHom** [comment]
Thanks for the request! :)

From how I understand your request, you want to be able to control what conditions should be checked for as a constant condition. In your case, if it's a constant condition that involves true or false, the`no-constant-condition` rule should not report an error. I think that given the different constant conditions, the option might be a white list option of what not to check imo.

Currently, the rule takes in the option `checkLoops`. If the option of checkLoops is:
1. true (default) with new option: loops with true or false would not be reported as errors; non-loops will be checked for constant condition (minus what's "set" with the new option)
2. false with new option: loops will not be checked, but the the optionality should be applied to non-loops.

With that said, I wonder if there's any existing rule that takes in a list of options like that. Or whether enough users would just want it for true and false as an exception. Hoping others can chime in too.

**Comment by guymguym** [comment]
Hi @VictorHom, Thanks for following up.

There are similar rules taking a list of exceptions for example `no-console` taking the option `{ "allow": ["error", "warn"] }` so I think it makes sense to support it similarly.

I think that a fixed option such as `{ "allowTrueAndFalse": true }` might be too rigid, so I would prefer `{ "allow": [true, false, 1, 0] }`.

Would be great to hear others preferences on this.

**Comment by not-an-aardvark** [comment]
> We do not want to disable that rule because it protects us from unintended code such as `while (x=1)` and similar error prone patterns.

Could you use [`no-cond-assign`](https://eslint.org/docs/rules/no-cond-assign) instead to avoid that type of code? It seems like it shouldn't be necessary to use `no-constant-condition` for that.

**Comment by guymguym** [comment]
@not-an-aardvark Thanks! That's a good suggestion.

There are still some other cases in [no-constant-condition](https://eslint.org/docs/rules/no-constant-condition) that we did not want to opt in for such as these:

Examples of incorrect code for this rule:
```
/*eslint no-constant-condition: "error"*/

if (false) {
    doSomethingUnfinished();
}

if (void x) {
    doSomethingUnfinished();
}

for (;-2;) {
    doSomethingForever();
}

while (typeof x) {
    doSomethingForever();
}

do {
    doSomethingForever();
} while (x = -1);

var result = 0 ? a : b;
```

And BTW I was surprised to see that there is a way to run infinite loops even with no-constant-condition using this syntax that is not detected by this rule:
```
for (;;) {
    doSomethingForever();
}
```
Any idea why is that behaving differently?
We might use it if we understand why this syntax is not considered a constant condition...

Thanks!

**Comment by SheetJSDev** [comment]
@guymguym there's a technical argument that `for(;;)` is using the implicit `true` behavior and therefore isn't explicitly a constant condition.

However, given that `while(true)` is a very common idiom among C expats, it seems strange that one should be deemed acceptable but not the other.  The spirit of the rule according to the documentation is:

> A constant expression (for example, a literal) as a test condition might be a typo or development trigger for a specific behavior. 

If the intention is to specifically permit `for(;;)` as a backdoor for nontrivial loops but not allow `while(true)`, it would be helpful to explicitly say it.  Merely putting the code in the documentation raises more questions than it answers.

Additionally, should there be an autofix to replace `while(true)` with `for(;;)`?

**Comment by nzakas** [comment]
Unfortunately, it looks like there wasn't enough interest from the team or community to implement this change. While we wish we'd be able to accommodate everyone's requests, we do need to prioritize. We've found that issues failing to be implemented after 90 days tend to never be implemented, and as such, we [close those issues](https://eslint.org/docs/maintainer-guide/issues#when-to-close-an-issue). This doesn't mean the idea isn't interesting or useful, just that it's not something the team can commit to.
